examples/slackbot/Notebook.ipynb

```diff
@@ -27,89 +27,93 @@
-    "    .count(window=duration(timedelta(minutes=10))),\n",
-    "  # The timestamp of the last message\n",
-    "  time=time_of(messages_by_channel),\n",
-    ")\n",
+    "        # We'll train ChatGPT to generate the user ID who will engage next\n",
+    "        completion=completion.col(\"user_id\"),\n",
+    "    )\n",
     "\n",
-    "# We'll train ChatGPT to generate the user ID who will engage next\n",
-    "completion = \"\\n\\n###\\n\\n\" + message.get(\"user_id\")\n",
+    "def main():\n",
+    "    # Initialize Kaskada with a local execution context.\n",
+    "    k.init_session()\n",
     "\n",
-    "# Each example will predict who will engage next, given\n",
-    "# the last 10 minutes of conversation history\n",
-    "examples = k.make_record(\n",
-    "  prompt=prompt.lag(1).lookup(message.get(\"channel\")),\n",
-    "  completion=completion,\n",
-    ")\n",
-    "examples = examples.when(examples.get(\"prompt\").is_not_null())\n",
+    "    tl = examples(messages = k.ParquetSource(files=[\"./messages.parquet\"]))\n",
+    "    tl = tl.when(tl.col(\"prompt\").is_valid())\n",
+    "    examples_df = tl.run().to_pandas()\n",
     "\n",
-    "# Split training & validation\n",
-    "from sklearn.model_selection import train_test_split\n",
-    "train, valid = train_test_split(\n",
-    "  examples.to_pandas(),\n",
-    "  test_size=0.2, random_state=42\n",
-    ")\n",
-    "train.to_json(\"train.jsonl\", orient='records', lines=True)\n",
-    "valid.to_json(\"valid.jsonl\", orient='records', lines=True)\n",
+    "    # Split training & validation\n",
     "\n",
-    "# Fine tune a model\n",
-    "!openai api fine_tunes.create -t \"train.jsonl\" -v \"valid.jsonl\"\n",
+    "    train, valid = train_test_split(examples_df, test_size=0.2, random_state=42)\n",
+    "    train.to_json(\"train.jsonl\", orient='records', lines=True)\n",
+    "    valid.to_json(\"valid.jsonl\", orient='records', lines=True)\n",
+    "\n",
+    "    # Fine tune a model\n",
+    "    !openai api fine_tunes.create -t \"train.jsonl\" -v \"valid.jsonl\"\n",
+    "\n",
+    "    # Apply in real-time\n",
+    "    live_messages = k.StreamSource()\n",
+    "\n",
+    "    # TODO: Spawn a Slack listener\n",
     "\n",
-    "# Apply in real-time\n",
-    "for p in prompt.to_generator(starting=datetime.now()):\n",
-    "  completions = openai.ChatCompletion.create(\n",
-    "    model=\"ft-2zaA7qi0rxJduWQpdvOvmGn3\", \n",
-    "    messages=[{\"role\": \"user\", \"content\": p}],\n",
-    "    n=5, logprobs=5,\n",
-    "  )\n",
-    "  for completion in completions:\n",
-    "    if sum(completion.logprobs) > 0.05:\n",
-    "      user = completion.completion\n",
-    "      # Use the Slack API to PM the user?\n",
-    "      print(f\"Notify {user} of conversation: \\n{p}\") "
+    "    # Handle messages in realtime\n",
+    "    for p in prompt(live_messages).run(starting=datetime.now()).to_generator():\n",
```

Comment:
Still TBD how feasible some of these in a short time frame.

Reply:
It's going to be hard to demo this if we can't do something in real time. Let's keep an eye on this - shortcuts / simplifications are fine.